src/db-adapters/Mongoose/MongooseAdapter.js

```diff
@@ -603,30 +604,40 @@ export default class MongooseAdapter {
     return words.join(" ");
   }
 
-  static getIdQueryType(idOrIds) {
-    const mode = typeof idOrIds === "string" ? "findOne" : "find";
-    let idQuery;
-
-    if (typeof idOrIds === "string") {
-      if (!this.idIsValid(idOrIds)) {
-        throw new APIError(404, undefined, "No matching resource found.", "Invalid ID.");
-      }
-
-      idQuery = {_id: idOrIds};
+  static getIdQueryType(idOrIds, Model) {
+    if (idOrIds == null) {
+      return Q([ "find", undefined ]);
     }
 
-    else if (Array.isArray(idOrIds)) {
-      if (!idOrIds.every(this.idIsValid)) {
-        throw new APIError(400, undefined, "Invalid ID.");
-      }
-
-      idQuery = {_id: {"$in": idOrIds}};
-    }
+    const [ mode, ids ] = Array.isArray(idOrIds) ? [ "find", idOrIds ] : [ "findOne", [ idOrIds ] ];
+    const idValidationPromises = ids.map(id => this.idIsValid(id, Model));
 
-    return [mode, idQuery];
+    return Q.all(idValidationPromises).then((idsAreValid) => {
```

Comment:
Ugh, now we're back to waiting for all the validations to finish before we run the `then` block, even if we could skip to throwing an error earlier. 

This is my fault... I said earlier that, if we're going to keep `isIdValid` named as it is, it should resolve with true or false. But, what I meant was that it should *resolve* with true if the id is valid but *reject* with false otherwise. That way, the basic structure of `getIdQueryType` can be as it was in my last comment.

I realize that rejecting with false is a bit weird semantically, though, which is why I edited my earlier comment to say that it's probably better to just rename `isIdValid`. So that's what I'd do now.

I.e., I'd rename `isIdValid` to `validateId`, which can then resolve/reject as you had originally (i.e., with the valid id on success, and the validation error on failure). And then the `getIdQueryType` can work with the two different handlers.